quickwit/quickwit-search/src/leaf.rs

```diff
@@ -1236,44 +1238,58 @@ pub async fn leaf_search(
             continue;
         }
 
-        leaf_search_single_split_futures.push(tokio::spawn(
-            leaf_search_single_split_wrapper(
-                request,
-                searcher_context.clone(),
-                index_storage.clone(),
-                doc_mapper.clone(),
-                split,
-                split_filter.clone(),
-                incremental_merge_collector.clone(),
-                leaf_split_search_permit,
-                aggregations_limits.clone(),
-            )
-            .in_current_span(),
+        leaf_search_single_split_join_handles.push((
+            split.split_id.clone(),
+            tokio::spawn(
+                leaf_search_single_split_wrapper(
+                    request,
+                    searcher_context.clone(),
+                    index_storage.clone(),
+                    doc_mapper.clone(),
+                    split,
+                    split_filter.clone(),
+                    incremental_merge_collector.clone(),
+                    leaf_split_search_permit,
+                    aggregations_limits.clone(),
+                )
+                .in_current_span(),
+            ),
         ));
     }
 
     // TODO we could cancel running splits when !run_all_splits and the running split can no
     // longer give better results after some other split answered.
-    let split_search_results: Vec<Result<(), _>> =
-        futures::future::join_all(leaf_search_single_split_futures).await;
+    let mut split_search_join_errors: Vec<(String, JoinError)> = Vec::new();
+
+    // There is no need to use `join_all`, as these are spawned tasks.
+    for (split, leaf_search_join_handle) in leaf_search_single_split_join_handles {
+        // splits that did not panic were already added to the collector
+        if let Err(join_error) = leaf_search_join_handle.await {
+            if join_error.is_cancelled() {
+                // An explicit task cancellation is not an error.
+                continue;
+            }
+            if join_error.is_panic() {
+                error!(split=%split, "leaf search task panicked");
+            } else {
+                error!(split=%split, "please report: leaf search was not cancelled, and could not extract panic. this should never happen");
+            }
+            split_search_join_errors.push((split, join_error));
```

Comment:
is there a reason to not call `incremental_merge_collector.add_failed_split()` directly from here?

Reply:
yes. We get the incremental_merge_collector by unwrapping the Arc.
The unwrap will only work if all the other references have been dropped. 

So this is loop is both to wait for all of the task to terminate (successfully or not), and to gather errors.